### PR TITLE
Cs 161 add message handler for invitation requests by calling api 

### PR DESCRIPTION
**Pull request by Pauljgraham**

### CODE REVIEW

File: src/SFA.DAS.Apprentice.LoginService.MessageHandler/SendInvitationCommandHandler.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                    ApprenticeshipName = sendInvitationCommand.ApprenticeshipName,
+                    Inviter = null
+                });
+
+                log.LogInformation(
+                    $"Completed {typeof(SendInvitationCommand)} InvitationId : {response?.InvitationId} Invited : {response?.Invited}, Message : { response?.Message}");
+            }
+            catch (Exception e)
+            {
+                log.LogError($"Errored when processing {typeof(SendInvitationCommand)}", e);
```
Comment (benagain) [inline src/SFA.DAS.Apprentice.LoginService.MessageHandler/SendInvitationCommandHandler.cs:50]: Should we re-throw the exception here as well?  That will let the messaging infrastructure perform its retries - if we swallow the exception then NSB will consider everything to have succeeded.

File: src/SFA.DAS.Apprentice.LoginService.MessageHandler/SendInvitationCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            }
+        }
+
+#if DEBUG
```
Comment (benagain) [inline src/SFA.DAS.Apprentice.LoginService.MessageHandler/SendInvitationCommandHandler.cs:54]: Good idea limiting the test trigger to debug only

File: src/SFA.DAS.LoginService.Application/Invitations/CreateInvitation/CreateInvitationHandler.cs [hunk 3]
```diff
@@ -55,8 +55,8 @@ public class CreateInvitationHandler : IRequestHandler<CreateInvitationRequest,
             }
             
             _logger.LogInformation($"CreateInvitationHandler : Client: {JsonConvert.SerializeObject(client)}");
-            
-            var existingUser = await _userService.FindByEmail(request.Email);
+
+            var existingUser = await _userAccountService.FindByEmail(request.Email);
```
Comment (benagain) [inline src/SFA.DAS.LoginService.Application/Invitations/CreateInvitation/CreateInvitationHandler.cs:59]: I'm curious about this change - does the `UserManager` (the backing behind the previous `UserService`) not contain the user?

File: src/SFA.DAS.LoginService.Application/Services/UserAccountService.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using SFA.DAS.LoginService.Application.Interfaces;
+using SFA.DAS.LoginService.Data;
+using SFA.DAS.LoginService.Data.Entities;
+
+namespace SFA.DAS.LoginService.Application.Services
+{
+    public class UserAccountService : IUserAccountService
```
Comment (benagain) [inline src/SFA.DAS.LoginService.Application/Services/UserAccountService.cs:10]: Maybe we can improve this name?  It doesn't speak to me why it is different to `UserService`.  

I'm afraid it's not clear to me why we're going directly to the database and bypassing the `UserManager`.

File: src/SFA.DAS.LoginService.Configuration/SFA.DAS.LoginService.Configuration.csproj [hunk 5]
```diff
@@ -5,7 +5,8 @@
     </PropertyGroup>
 
     <ItemGroup>
-      <PackageReference Include="Newtonsoft.Json" Version="12.0.1" />
+      <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
```
Comment (benagain) [inline src/SFA.DAS.LoginService.Configuration/SFA.DAS.LoginService.Configuration.csproj:8]: Annoying Newtonsoft horrible dependecy problems

File: src/SFA.DAS.Apprentice.LoginService.MessageHandler.AcceptanceTests/Bindings/MessageBus.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        public MessageBus(TestContext context)
+        {
+            _context = context;
+        }
+
+        [BeforeScenario(Order = 2)]
+        public async Task InitialiseFunctions()
+        {
+            _context.TestDirectory = new DirectoryInfo(Path.Combine(Directory.GetCurrentDirectory(), Guid.NewGuid().ToString()));
```
[the remark this answers is not in the document]
Comment (Pauljgraham) [reply src/SFA.DAS.Apprentice.LoginService.MessageHandler.AcceptanceTests/Bindings/MessageBus.cs:24]: yes, good idea. But I'm also considering moving the location of this to a lower directory as it breaks the Accept Tests project when it fails and the directory  isn't deleted

File: src/SFA.DAS.Apprentice.LoginService.MessageHandler/SendInvitationCommandHandler.cs (hunk 1 above)
Comment (Pauljgraham) [reply src/SFA.DAS.Apprentice.LoginService.MessageHandler/SendInvitationCommandHandler.cs:50]: Good spot, missed that

File: src/SFA.DAS.LoginService.Application/Invitations/CreateInvitation/CreateInvitationHandler.cs (hunk 3 above)
Comment (Pauljgraham) [reply src/SFA.DAS.LoginService.Application/Invitations/CreateInvitation/CreateInvitationHandler.cs:59]: UserManager contains HttpContext and isn't available with a message handler. So this rework allows us to call this application layer from the web or the MessageHandler

File: src/SFA.DAS.LoginService.Application/Services/UserAccountService.cs (hunk 4 above)
Comment (Pauljgraham) [reply src/SFA.DAS.LoginService.Application/Services/UserAccountService.cs:10]: I agree, just not sure what to name it

File: src/SFA.DAS.LoginService.Configuration/SFA.DAS.LoginService.Configuration.csproj (hunk 5 above)
Comment (Pauljgraham) [reply src/SFA.DAS.LoginService.Configuration/SFA.DAS.LoginService.Configuration.csproj:8]: yes, I'm sure it doesn't matter but the DAS NSB package insists, and I don't know why

File: src/SFA.DAS.LoginService.Application/Services/UserAccountService.cs (hunk 4 above)
Comment (Pauljgraham) [reply src/SFA.DAS.LoginService.Application/Services/UserAccountService.cs:10]: We could call it 'WithoutHttpContextUserService' I know it sounds like a joke but it's purpose becomes clear

Comment (benagain) [reply src/SFA.DAS.LoginService.Application/Services/UserAccountService.cs:10]: Ok, I see the reason now.

Hmm, how about switching it round and calling the class without HttpContext `UserManager` and the original one `HttpContextUserManager` or something like that?

**Review by benagain — APPROVED** [review]

**Review by arvinthseran — APPROVED** [review]

**Pull request merged**